crates/p2p/src/main_loop.rs

```diff
@@ -171,6 +178,49 @@ impl MainLoop {
             SwarmEvent::ConnectionEstablished {
                 peer_id, endpoint, ..
             } => {
+                // If this is an incoming connection, we have to prevent the peer from
+                // reconnecting too quickly.
+                if endpoint.is_listener() {
+                    // Extract the IP address of the peer from his multiaddr.
+                    let ip = endpoint.get_remote_address().iter().find_map(|p| match p {
+                        Protocol::Ip4(ip) => Some(IpAddr::V4(ip)),
+                        Protocol::Ip6(ip) => Some(IpAddr::V6(ip)),
+                        _ => None,
+                    });
+                    match ip {
+                        Some(ip) => {
+                            // If the peer is in the recent peers set, this means he is attempting to
+                            // reconnect too quickly. Close the connection.
+                            if self.recent_peers.read().await.contains(&ip) {
+                                tracing::debug!(%peer_id, "Peer attempted to reconnect too quickly, closing");
+                                if let Err(e) = self.disconnect(peer_id).await {
+                                    tracing::error!(%e, "Failed to disconnect peer");
+                                }
+                                return;
+                            } else {
+                                // Otherwise, add the peer to the recent peers set. After a certain period,
+                                // remove the peer from the recent peers set. This allows us to prevent
+                                // malicious peers from reconnecting too quickly.
+                                let recent_peers = self.recent_peers.clone();
+                                let timeout = self.connection_timeout;
+                                recent_peers.write().await.insert(ip);
+                                tokio::spawn(async move {
+                                    tokio::time::sleep(timeout).await;
+                                    recent_peers.write().await.remove(&ip);
+                                });
```

Comment:
We could maybe used a [TimedSizedCache](https://docs.rs/cached/latest/cached/stores/struct.TimedSizedCache.html) instead of spawning a task per IP address? There's nothing limiting how many tasks we're spawning here so I feel this would be fairly easy to DoS?

Reply:
Or [HashSetDelay](https://docs.rs/delay_map/latest/delay_map/hashset_delay/struct.HashSetDelay.html) if we need a set. My point is that we probably shouldn't spawn a task per IP just to remove the item from the set.